torch/csrc/jit/script/compilation_unit.h

```diff
@@ -137,17 +137,27 @@ struct TORCH_API CompilationUnit {
   /**
    * Register a class as being owned by this compilation unit.
    */
-  void register_class(c10::NamedTypePtr classType) {
-    classes_.push_back(std::move(classType));
+  void register_class(c10::NamedTypePtr namedType) {
+    if (auto classType = namedType->cast<c10::ClassType>()) {
+      // TODO: class types cannot be redefined because we have no way right now
```

Comment:
yuck -- we should have consistent behavior. I don't think we should let anything get redefined.

Reply:
Definitely agree this is gross, but we're in a difficult spot. Basically: 
- Our tests frequently redefine Functions. That worked before because they all went in a different CU, but now there is a global one.
- But aren't these functions all different because they occur nested in different functions? No, because in Python there is no way of knowing whether a function is nested within another.
- Okay, fine. But then shouldn't we clear the Python CU between every test for isolation anyway? But we can't do that, because we define script funcs in the standard library and the resolvers for our methods will capture a dangling StrongFunctionPtr.

I think the last one can be untangled, but I didn't want to couple it with this PR. Also, in the long run, do we really want the behavior to be inconsistent with Python (where you can redefine pretty much anything)?

Reply:
There is a way to figure out whether something is nested or not: get its qualified name, try to access it starting from the global environment, did you get the function again? If so, it has that global name, if not, it was nested in something.  Global names are important for export/import but should never be used as the way we link functions together in the Python environment. Given that we do not consistently keep track of when we inline, it will result in all sorts of bugs. If something is not a globally exposed function (i.e. accessible in python following the module hierarchy), we should be comfortable giving it a mangled name.

Reply:
Ah I see what you mean now. As an aside, one issue I'm running into with what you proposed is that when we do:
```
@torch.jit.script
def foo():
    ...
```
The `script()` decorator cannot figure out whether it can access `foo` from the global environment, because while `script()` is running `foo` has not been defined yet. So we can't, at definition time, let the `CU` know whether the function name needs to be mangled or not.

Let me know if you have any ideas. Possibly there is some Python trick that I need to Google for.